routing/index_graph_starter.hpp

```diff
@@ -21,136 +26,180 @@ class IndexGraphStarter final
   using TEdgeType = IndexGraph::TEdgeType;
   using TWeightType = IndexGraph::TWeightType;
 
-  class FakeVertex final
+  struct FakeEnding
   {
-  public:
-    FakeVertex(NumMwmId mwmId, uint32_t featureId, uint32_t segmentIdx, m2::PointD const & point)
-      : m_segment(mwmId, featureId, segmentIdx, true /* forward */)
-      , m_junction(point, feature::kDefaultAltitudeMeters)
-    {
-    }
-
-    FakeVertex(Segment const & segment, m2::PointD const & point, bool strictForward)
-      : m_segment(segment)
-      , m_junction(point, feature::kDefaultAltitudeMeters)
-      , m_strictForward(strictForward)
-    {
-    }
-
-    FakeVertex(Segment const & segment, Junction const & junction, bool strictForward)
-      : m_segment(segment), m_junction(junction), m_strictForward(strictForward)
-    {
-    }
-
-    NumMwmId GetMwmId() const { return m_segment.GetMwmId(); }
-    uint32_t GetFeatureId() const { return m_segment.GetFeatureId(); }
-    Junction const & GetJunction() const { return m_junction; }
-    m2::PointD const & GetPoint() const { return m_junction.GetPoint(); }
-    Segment const & GetSegment() const { return m_segment; }
-
-    Segment GetSegmentWithDirection(bool forward) const
-    {
-      return Segment(m_segment.GetMwmId(), m_segment.GetFeatureId(), m_segment.GetSegmentIdx(),
-                     forward);
-    }
-
-    bool Fits(Segment const & segment) const
-    {
-      return segment.GetMwmId() == m_segment.GetMwmId() &&
-             segment.GetFeatureId() == m_segment.GetFeatureId() &&
-             segment.GetSegmentIdx() == m_segment.GetSegmentIdx() &&
-             (!m_strictForward || segment.IsForward() == m_segment.IsForward());
-    }
-
-    uint32_t GetSegmentIdxForTesting() const { return m_segment.GetSegmentIdx(); }
-    bool GetStrictForward() const { return m_strictForward; }
-
-  private:
-    Segment m_segment;
-    Junction m_junction;
-    // This flag specifies which fake edges should be placed from the fake vertex.
-    // true: place exactly one fake edge to the m_segment with indicated m_forward.
-    // false: place two fake edges to the m_segment with both directions.
-    bool m_strictForward = false;
+    Junction m_originJunction;
+    std::vector<Junction> m_projectionJunctions;
```

Comment:
If you keeping the invariant that m_projectionJunction.size() is the same as m_projectionSegments.size(), why not to replace a pair of vectors by a vector of pairs? Or, even better, vector of structs with named fields.

Reply:
done